Propose underscores in numeric literals

takenobu-hs opened the pull request:
This is a proposal to add underscores to numeric literals.
Underscores (`_`) in numeric literals are simply ignored.

[Rendered](https://github.com/takenobu-hs/ghc-proposals/blob/numeric-underscores/proposals/0000-numeric-underscores.rst)

nomeata commented on proposals/0000-numeric-underscores.rst:
You should probably make the proposal right away suggest this language extension, as a language extensions is required for such changes.

bgamari replied:
We could fold it into `BinaryLiterals` or something like this, but yes, I generally agree that a new extension is probably best.

takenobu-hs replied:
Thanks for the feedback.
I agree with the language extension.
I will update the proposal for specifying language extension.

takenobu-hs replied:
I explicitly described the language extension at [this commit][1].

[1]: https://github.com/ghc-proposals/ghc-proposals/pull/76/commits/ca707d6fb8b3d57c5307283a00c00af6f4cd1d62

cartazio commented:
I support this proposal.
There’s a bunch of details to make it work nicely but it sounds like
everyone’s got those well In hand.

Heck, do we have a candidate patch yet? I might try if there’s not one
already!

On Sat, Sep 30, 2017 at 10:46 AM Ben Gamari <notifications@github.com>
wrote:

> *@bgamari* commented on this pull request.
> ------------------------------
>
> In proposals/0000-numeric-underscores.rst
> <https://github.com/ghc-proposals/ghc-proposals/pull/76#discussion_r142007591>
> :
>
> > +    isUnderMillion = (< 1_000_000)
> +
> +    clip64M x
> +        | x > 0x3ff_ffff = 0x3ff_ffff
> +        | otherwise = x
> +
> +    test8bit x = (0b01_0000_0000 .&. x) /= 0
> +
> +Effect and Interactions
> +-----------------------
> +I believe that this proposal will improve the readability, quality and expressiveness of native numeric literals without degrading performance.
> +
> +Costs and Drawbacks
> +-------------------
> +* Implementation costs are mostly related to lexers.
> +* Maintenance costs are related to compatibility. Compatibility can be handled with language extension like ``NumericUnderscores``.
>
> We could fold it into BinaryLiterals or something like this, but yes, I
> generally agree that a new extension is probably best.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ghc-proposals/ghc-proposals/pull/76#discussion_r142007591>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAQwiqeBleqj6HbjTd6CsJpaSE-l52jks5snlQ9gaJpZM4PpdoI>
> .
>

phadej commented on proposals/0000-numeric-underscores.rst:
FWIW, for primitive types, like `Int` GHC does constant fold the expression.

takenobu-hs replied:
Thank you for helpful information.

phadej commented on proposals/0000-numeric-underscores.rst:
This would allow trailing underscore, i.e. `1000_1000_`, do we want such thing? If we do, then there should be an example in proposal.

takenobu-hs replied:
Thank you.
For float only, I assumed the trailing underscore.

For example, the following description is possible.
  gravity = 6.674_08_e−11

cartazio commented:
i'd personally thing we dont want trailing underscores, but i'm open to
being convinc

On Sat, Sep 30, 2017 at 12:51 PM, Oleg Grenrus <notifications@github.com>
wrote:

> *@phadej* commented on this pull request.
> ------------------------------
>
> In proposals/0000-numeric-underscores.rst
> <https://github.com/ghc-proposals/ghc-proposals/pull/76#discussion_r142010106>
> :
>
> > +I believe that this proposal will improve the readability, quality and expressiveness of native numeric literals without degrading performance.
> +
> +Costs and Drawbacks
> +-------------------
> +* Implementation costs are mostly related to lexers.
> +* Maintenance costs are related to compatibility. Compatibility can be handled with language extension like ``NumericUnderscores``.
> +* I think the user's learning curve is not a problem. They will soon get used to it.
> +* Syntax highlighting for text editors and code browsers is affected.
> +
> +Alternatives
> +------------
> +For example, these expressions are current alternatives:
> +
> +.. code-block:: none
> +
> +    x = 10 * 000 * 000 :: Int
>
> FWIW, for primitive types, like Int GHC does constant fold the expression.
> ------------------------------
>
> In proposals/0000-numeric-underscores.rst
> <https://github.com/ghc-proposals/ghc-proposals/pull/76#discussion_r142010132>
> :
>
> > +I propose an extension to the existing syntax of numeric literals.
> +
> +Current syntax:
> +
> +.. code-block:: none
> +
> +    decimal     →  digit{digit}
> +    octal       →  octit{octit}
> +    hexadecimal →  hexit{hexit}
> +    binary      →  binit{binit}
> +
> +New syntax (this proposal):
> +
> +.. code-block:: none
> +
> +    decimal     →  digit{[_ | digit]}
>
> This would allow trailing underscore, i.e. 1000_1000_, do we want such
> thing? If we do, then there should be an example in proposal.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ghc-proposals/ghc-proposals/pull/76#pullrequestreview-66317259>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAQwvJMMzZ2SZefNFFCSHH1SovQ03bPks5snnGEgaJpZM4PpdoI>
> .
>

takenobu-hs commented:
> Heck, do we have a candidate patch yet? I might try if there’s not one already!

@cartazio, Thank you, I'm glad.
I am very glad if you do it!

takenobu-hs commented:
> i'd personally thing we dont want trailing underscores, but i'm open to being convinc

phadej is also mentioned.
Please give me feedback on the necessity of trailing underscore.
The specification definition may be a little complicated.

takenobu-hs commented:
I studied alternative proposal for the trailing underscore.
Please give me feedback which is better.

(1) current proposal:

    decimal     →  digit{[_ | digit]}
    octal       →  octit{[_ | octit]}
    hexadecimal →  hexit{[_ | hexit]}
    binary      →  binit{[_ | binit]}

    exponent → (e | E) [+ | -] decimal


(2) alternative proposal for trailing underscore:

    decimal     →  digit[{[_ | digit]} digit]
    octal       →  octit[{[_ | octit]} octit]
    hexadecimal →  hexit[{[_ | hexit]} hexit]
    binary      →  binit[{[_ | binit]} binit]

    exponent → [_] (e | E) [+ | -] decimal


Examples with (2) alternative proposal:

    x0 = 1_000_000   -- valid
    x1 = 1000000_    -- invalid
    x2 = _1000000    -- invalid

    e0 = 0.0001      -- valid
    e1 = 0.000_1     -- valid
    e2 = 0_.0001     -- invalid
    e3 = _0.0001     -- invalid
    e4 = 0._0001     -- invalid
    e5 = 0.0001_     -- invalid

    f0 = 1e+23       -- valid
    f1 = 1_e+23      -- valid
    f2 = 1e_+23      -- invalid

    g0 = 1e+23       -- valid
    g1 = 1e+_23      -- invalid
    g2 = 1e+23_      -- invalid

osa1 commented:
Strong :+1: from me. We often do things like

```
1 * 1000 * 1000
threadDelay 5000000 -- five sec
```

first one is weird (GHC can evaluate that in compile time but still), second one is dangerous if you forget to update the commend after updating the code. Rust has this feature and I often use it.

treeowl commented:
The first "alternative" currently has multiplication by zero, which looks like a mistake. I'm generally in favor of the proposal. I don't think trailing underscores are useful. Furthermore, leading underscores clash with numbered holes.

```haskell
f x = _1
```

Currently, that is a typed hole named `_1`. I'd I understand the proposal correctly, it would become the number 1. Not a disaster by any means, but also not what I want. In partial type signatures, that could even break working code:

```haskell
foo :: _1 -> Int
```

would now produce a kind error.

treeowl commented:
Previous comment edited to remove bogus information. Sorry for the confusion.

osa1 commented:
Agreed that numerals with leading underscore should not be considered as a number. Trailing underscores don't seem useful but at least harmless.

treeowl commented:
I misread the proposal before; it seems leading underscores are already
excluded.

On Oct 2, 2017 2:26 AM, "Ömer Sinan Ağacan" <notifications@github.com>
wrote:

> Agreed that numerals with leading underscore should not be considered as a
> number. Trailing underscores don't seem useful but at least harmless.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ghc-proposals/ghc-proposals/pull/76#issuecomment-333450202>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABzi_f31-mO6ud8vGDFqA0pvFEAHO9Pxks5soIIZgaJpZM4PpdoI>
> .
>

simonpj commented:
I'm happy with this.

merijn commented:
I'm strongly +1 on this too.

takenobu-hs commented:
Hi everyone, I'm glad responses! Thank you very much.
I updated proposal:
  * Modify specification to remove trailing underscores
  * Add validity examples
  * Correct and add alternative examples


I chose the option to remove the trailing underscore.
I think that is easier to recognize as a number.

Please tell me if something is wrong.

takenobu-hs commented:
I will wait for a few weeks of discussion period :)

By the way, I wrote a patch for my studies [1].
However, I am glad that Carter or someone will write a more suitable patch.
Especially, it would be nice if someone could express Lexer simpler.

[1] https://github.com/ghc/ghc/compare/master...takenobu-hs:wip/numeric-underscores

cartazio commented:
woah, awesome!

i got a bit sick for a week + buried in mentoring a colleague, so i've not
yet had a chance to dig into this properly, :/

i will endeavor to spend some time this weekend/next week to review both
your draft patch and the current state of affairs in the doc :)

On Sat, Oct 14, 2017 at 5:25 AM, Takenobu Tani <notifications@github.com>
wrote:

> I will wait for a few weeks of discussion period :)
>
> By the way, I wrote a patch for my studies [1].
> However, I am glad that Carter or someone will write a more suitable patch.
> Especially, it would be nice if someone could express Lexer simpler.
>
> [1] ghc/ghc@master...takenobu-hs:wip/numeric-underscores
> <https://github.com/ghc/ghc/compare/master...takenobu-hs:wip/numeric-underscores>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ghc-proposals/ghc-proposals/pull/76#issuecomment-336622923>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAQwtyA_l5WQoUqUSQthvMMzvVBMkd7ks5ssH4fgaJpZM4PpdoI>
> .
>

takenobu-hs commented:
@cartazio thank you, I am grateful for your help :)
I am not in a hurry, so please do it when you have time.
Please take care of yourself.

cartazio commented:
I did a first read of your patch.
One question is: how do we test for failures / disallowed patterns ?

On Fri, Oct 20, 2017 at 8:02 AM Takenobu Tani <notifications@github.com>
wrote:

> @cartazio <https://github.com/cartazio> thank you, I am grateful for your
> help :)
> I am not in a hurry, so please do it when you have time.
> Please take care of yourself.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ghc-proposals/ghc-proposals/pull/76#issuecomment-338187112>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAQwil5EpMfK9OOgAu74l8NLLUzP33Rks5suIvQgaJpZM4PpdoI>
> .
>

takenobu-hs commented:
Thank you, I will add a testcase for failures / disallowed patterns in `testsuite/tests/parser/should_fail/` directory.

takenobu-hs commented:
Hi, I added the testcase in this [1].
(testsuite/tests/parser/should_fail/NumericUnderscoresFail0.hs)
Please let me know if you have any problems. I am not in a hurry. Thanks.

[1] https://github.com/ghc/ghc/compare/master...takenobu-hs:wip/numeric-underscores

takenobu-hs commented:
Four weeks have passed since I submitted this proposal.
I think the discussion has converged.
I would like to submit this for a decision.

cc @nomeata

nomeata commented:
Committee decision process started. Thanks for the ping, @bitemyapp, I missed the earlier one among too many GitHub notifications.

takenobu-hs commented:
@nomeata, thank you very much :)

cartazio commented:
This or some honed refinement is also likely merited for Inclusion in H2020

@takenobu could you explain the qualms you mentioned earlier about the
lexer changes you wrote? I must confess I didn’t quite understand it at the
Time :)

On Thu, Nov 2, 2017 at 8:35 AM Takenobu Tani <notifications@github.com>
wrote:

> @nomeata <https://github.com/nomeata>, thank you very much :)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ghc-proposals/ghc-proposals/pull/76#issuecomment-341407485>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAQwkxE7XPlXDFat0vCfOXB1Jdan4wKks5sybb2gaJpZM4PpdoI>
> .
>

takenobu-hs commented:
@cartazio thank you always.
If this proposal is accepted, I will send the patch to phab.
I would be happy if you could give me comments there :)

yav commented:
I think we should allow underscores between the base specifier and the digits for binary, octal and hex literals.   For example, I think it would be handy to write   `0x_FFFF_FFFF`   instead of `0xFFFF_FFFF`

takenobu-hs commented:
@yav thanks for the advice.
I considered it as follows.

Current syntax (Haskell Report 2010 + BinaryLiterals + HexFloatLiterals):
```
decimal     →  digit{digit}
octal       →  octit{octit}
hexadecimal →  hexit{hexit}
binary      →  binit{binit}                                  -- BinaryLiterals

integer →  decimal
         | 0 (o | O) octal
         | 0 (x | X) hexadecimal
         | 0 (b | B) binary                                  -- BinaryLiterals

float →  decimal . decimal [exponent]
       | decimal exponent
       | 0 (x | X) hexadecimal . hexadecimal [bin_exponent]  -- HexFloatLiterals
       | 0 (x | X) hexadecimal bin_exponent                  -- HexFloatLiterals

exponent     →  (e | E) [+ | -] decimal
bin_exponent →  (p | P) [+ | -] decimal                      -- HexFloatLiterals
```

This proposal (revised):
```
decimal     →  digit[{_ | digit} digit]
octal       →  octit[{_ | octit} octit]
hexadecimal →  hexit[{_ | hexit} hexit]
binary      →  binit[{_ | binit} binit]

integer →  decimal
         | 0 (o | O) [_] octal                                      -- *** for 0o_123
         | 0 (x | X) [_] hexadecimal                                -- *** for 0x_ffff
         | 0 (b | B) [_] binary                                     -- *** for 0b_0101

float →  decimal . decimal [_] [exponent]
       | decimal [_] exponent
       | 0 (x | X) [_] hexadecimal . hexadecimal [[_]bin_exponent]  -- *** for 0x_ff.ff
       | 0 (x | X) [_] hexadecimal [_] bin_exponent                 -- *** for 0x_ffp1
```

The lines of `***` are the parts you pointed out.
Please tell me if there is any misunderstanding.
If this is correct, I will update the proposal file.

cartazio commented:
Possibly expositional question :
Would the spec change be cleaner / easier to write if we factor it into
“here’s all the valid liberals sans underscore” plus “here’s the extended
space of literal we also accept”?

I’m totally uncertain if my suggestion makes sense or  just adds complexity
:)

On Wed, Nov 8, 2017 at 7:11 AM Takenobu Tani <notifications@github.com>
wrote:

> @yav <https://github.com/yav> thanks for the advice.
> I considered it as follows.
>
> Current syntax (Haskell Report 2010 + BinaryLiterals + HexFloatLiterals):
>
> decimal     →  digit{digit}
> octal       →  octit{octit}
> hexadecimal →  hexit{hexit}
> binary      →  binit{binit}                                  -- BinaryLiterals
>
> integer →  decimal
>          | 0 (o | O) octal
>          | 0 (x | X) hexadecimal
>          | 0 (b | B) binary                                  -- BinaryLiterals
>
> float →  decimal . decimal [exponent]
>        | decimal exponent
>        | 0 (x | X) hexadecimal . hexadecimal [bin_exponent]  -- HexFloatLiterals
>        | 0 (x | X) hexadecimal bin_exponent                  -- HexFloatLiterals
>
> exponent     →  (e | E) [+ | -] decimal
> bin_exponent →  (p | P) [+ | -] decimal                      -- HexFloatLiterals
>
> This proposal (revised):
>
> decimal     →  digit[{_ | digit} digit]
> octal       →  octit[{_ | octit} octit]
> hexadecimal →  hexit[{_ | hexit} hexit]
> binary      →  binit[{_ | binit} binit]
>
> integer →  decimal
>          | 0 (o | O) [_] octal                                      -- *** for 0o_123
>          | 0 (x | X) [_] hexadecimal                                -- *** for 0x_ffff
>          | 0 (b | B) [_] binary                                     -- *** for 0b_0101
>
> float →  decimal . decimal [_] [exponent]
>        | decimal [_] exponent
>        | 0 (x | X) [_] hexadecimal . hexadecimal [[_]bin_exponent]  -- *** for 0x_ff.ff
>        | 0 (x | X) [_] hexadecimal [_] bin_exponent                 -- *** for 0x_ffp1
>
> The lines of *** are the parts you pointed out.
> Please tell me if there is any misunderstanding.
> If this is correct, I will update the proposal file.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ghc-proposals/ghc-proposals/pull/76#issuecomment-342799651>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAQwioPIyAl_yqmIk6-o6tRNma5x0mTks5s0ZqMgaJpZM4PpdoI>
> .
>

yav commented:
@takenobu-hs Yep, that's what I was thinking.

yav commented:
@takenobu-hs Yep, that's what I was thinking.

takenobu-hs commented:
@yav thanks for confirmation.

@cartazio thanks for a clear perspective.
I investigated to make it easy to understand the changed part of the specification.
I revise the proposal as follows. The part of `numSpacer` are the change points in this proposal.

This proposal (revised):
```
-- places affected by this proposal
numSpacer = _

-- underscores between digits
decimal     →  digit[{numSpacer | digit} digit]
octal       →  octit[{numSpacer | octit} octit]
hexadecimal →  hexit[{numSpacer | hexit} hexit]
binary      →  binit[{numSpacer | binit} binit]

-- an underscore succeeding the base specifier
integer →  decimal
         | 0 (o | O) [numSpacer] octal
         | 0 (x | X) [numSpacer] hexadecimal
         | 0 (b | B) [numSpacer] binary

float →  decimal . decimal [exponent]
       | decimal exponent
       | 0 (x | X) [numSpacer] hexadecimal . hexadecimal [bin_exponent]
       | 0 (x | X) [numSpacer] hexadecimal bin_exponent

-- an underscore preceding the exponent
exponent     →  [numSpacer] (e | E) [+ | -] decimal
bin_exponent →  [numSpacer] (p | P) [+ | -] decimal
```

Please let me know if it is insufficient.
If this is sufficient, I will update the proposal file.

yav commented:
@takenobu-hs I think we should allow multiple underscores next to each other, and at least one of your examples in the proposal suggests that you also think so.  To accommodate that I think that `numSpacer` should be `_+` (i.e. 1 or more underscores).

Also, would you mind updating the actual proposal with the current version of the syntax, so that we have the whole proposal in a single place.  

Here's another suggestion:  I would drop the "Current specification of numeric literals" section.  It is enough to just have "current syntax" and "new syntax",   or even just "new syntax".

takenobu-hs commented:
@yav thanks for the review.

> I think we should allow multiple underscores next to each other, and at least one of your examples in the proposal suggests that you also think so. To accommodate that I think that numSpacer should be _+ (i.e. 1 or more underscores).

In the current proposal, we could allow multiple underscores in decimal, octal, hexadecimal, and binary.
Because `{numSpacer | digit}` represents zero or more repetitions in Haskell Report 2010.
For example `decimal` allows `5__6`.

In addition, I changed the underscore before the exponent and after the base specifier to multiple in this revision.

> Also, would you mind updating the actual proposal with the current version of the syntax, so that we have the whole proposal in a single place.

Thanks. I pushed the actual proposal with the current version of the syntax.


> Here's another suggestion: I would drop the "Current specification of numeric literals" section. It is enough to just have "current syntax" and "new syntax", or even just "new syntax".

Thanks. That makes the proposal easier to understand.
I dropped the "Current specification of numeric literals" section.  And I reconstructed the specification part.

Please review the revised version.

takenobu-hs commented:
@yav I updated `numSpacer` to make regular expressions more uniform and simple.
I think that the changes from the current specification became easier to understand.
Please review this revised version.

yav commented:
Ok, I've marked this as accepted. 

@takenobu-hs you should consider submitting your patch to phabricator for review.   I had a quick look at it and I have just one major suggestion:   instead of having two sets of definitions for the literals (one with underscores and one without), it might be simpler to have only the definitions _with_ underscores, and then have a separate function that validates the literals, ensuring that there are no underscores unless the extension is enabled.

takenobu-hs commented:
Hi everyone thanks for a lot of advice and comments.
@yav, @nomeata, thanks for accept and merge. 

> @takenobu-hs you should consider submitting your patch to phabricator for review.

I will submit the patch to phabricator after improving the following.

> I had a quick look at it and I have just one major suggestion: instead of having two sets of definitions for the literals (one with underscores and one without),

Thanks for valuable suggestions. That was my concern.

> it might be simpler to have only the definitions with underscores, and then have a separate function that validates the literals, ensuring that there are no underscores unless the extension is enabled.

It's beautiful. I like your unified definition.
I will investigate how to implement the function for validation.
I have not understood the implementation manner in Lexer.x yet.
I'd be nice if someone tell me hints on how to implement it on trac or ghc-devs ML :)

By the way, should I create a ticket for this proposal?

nomeata commented:
> By the way, should I create a ticket for this proposal?

Sure, that’d be helpful.

takenobu-hs commented:
I created the ticket (https://ghc.haskell.org/trac/ghc/ticket/14473).

In addition, I added the ticket number to `0009-numeric-underscores.rst` and sent a PR (https://github.com/ghc-proposals/ghc-proposals/pull/95).

Thanks.
